INT-4361: Add a role() hook to Java DSL

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4361

* To get a gain of the method chain code flow and avoid extra annotation
parsing, add `.role(String)` hook to the `EndpointSpec`
* Delegate the provided `role` property to the `AbstractEndpoint`
* Register `AbstractEndpoint` as itself `SmartLifecycle` in the
`SmartLifecycleRoleController`
* Add `destroy()` to the `AbstractEndpoint` and remove it from the
`SmartLifecycleRoleController`
* Provide some Java 8 code style refactoring

artembilan committed:
INT-4361: Add a role() hook to Java DSL

JIRA: https://jira.spring.io/browse/INT-4361

* To get a gain of the method chain code flow and avoid extra annotation
parsing, add `.role(String)` hook to the `EndpointSpec`
* Delegate the provided `role` property to the `AbstractEndpoint`
* Register `AbstractEndpoint` as itself `SmartLifecycle` in the
`SmartLifecycleRoleController`
* Add `destroy()` to the `AbstractEndpoint` and remove it from the
`SmartLifecycleRoleController`
* Provide some Java 8 code style refactoring

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractEndpoint.java:
```diff
@@ -84,6 +108,21 @@ protected void onInit() throws Exception {
+					.getBean(IntegrationContextUtils.INTEGRATION_LIFECYCLE_ROLE_CONTROLLER,
+					SmartLifecycleRoleController.class);
+
+			this.roleController.addLifecycleToRole(this.role, this);
```
Needs null check - also, if endpoints now register themselves, shouldn't we use the parsers to use the same technique? https://github.com/spring-projects/spring-integration/pull/1483/files#diff-1fda9b32dc8bd19e9984d90ebf376ef6

artembilan replied:
Yeah... We have to catch:
```
catch (NoSuchBeanDefinitionException e) {
		this.logger.error("No LifecycleRoleController in the context");
}
```

Well, we can't use `addLifecycleToRole()` from the XML parser since we don't have bean instances there.
The manipulation with the `BeanDefinition` attributes is correct there.
IMO...

garyrussell replied:
? Can't we just set the `role` property on the endpoint BD from the xml?

artembilan committed:
* Rework XML parsers and Annotation processors to populate `role`
property on the `AbstractEndpoint`
* Wrap `roleController` bean extraction to the `NoSuchBeanDefinitionException` catch
* Fix several `AbstractEndpoint` implementation to properly call
`super.onInit()` which has been missed before

artembilan commented:
@garyrussell ,

Please, take a look.

That indeed was good suggestion to go.
This way I have observed several `AbstractEndpoint` implemetation which didn't delegate to `super.onInit()`.

I guess we need to back port that fix to `4.3.x` 😄

garyrussell commented:
Merged as f7e75223c79996e1350c87f35724491adfe6b39b
